source/s3_meta_request.c

```diff
@@ -2155,6 +2169,158 @@ void aws_s3_meta_request_result_setup(
+        aws_future_void_set_result(write_future);
+    }
+
+    if (illegal_usage_terminate_meta_request) {
+        aws_future_void_set_error(write_future, AWS_ERROR_INVALID_STATE);
+        aws_s3_meta_request_set_fail_synced(meta_request, NULL, AWS_ERROR_INVALID_STATE);
+    }
+
+    aws_s3_meta_request_unlock_synced_data(meta_request);
+    /* END CRITICAL SECTION */
+
+    if (ready_to_send || illegal_usage_terminate_meta_request) {
+        /* Schedule the work task, to continue processing the meta-request */
+        aws_s3_client_schedule_process_work(meta_request->client);
+    }
+
+    return write_future;
+}
+
+/* Copy pending async-write data into the buffer.
+ * This is only called when there's enough data for the next part. */
+static int s_s3_meta_request_read_from_pending_async_writes(
+    struct aws_s3_meta_request *meta_request,
+    struct aws_byte_buf *dest,
+    bool *eof) {
+
+    *eof = false;
+
+    struct aws_future_void *write_future_to_complete = NULL;
+
+    /* This stream-read should NOT happen unless there's a pending async-write */
+    AWS_ASSERT(meta_request->async_write.synced_future != NULL);
+
+    /* Buffered data should not exceed part-size */
+    AWS_ASSERT(dest->capacity - dest->len >= meta_request->async_write.buffered_data.len);
+
+    /* Copy all buffered data */
+    aws_byte_buf_write_from_whole_buffer(dest, meta_request->async_write.buffered_data);
+    meta_request->async_write.buffered_data.len = 0;
+
+    /* Copy as much unbuffered data as possible */
+    aws_byte_buf_write_to_capacity(dest, &meta_request->async_write.unbuffered_cursor);
+
+    /* We should have filled the dest buffer, unless this is the final write */
+    AWS_ASSERT(dest->len == dest->capacity || meta_request->async_write.eof);
+
+    /* If haven't received EOF, and there's not enough data in unbuffered_cursor to fill another part,
+     * then we need to move it into buffered_data, so we can complete the write's future and get more data */
+    if (!meta_request->async_write.eof && meta_request->async_write.unbuffered_cursor.len < meta_request->part_size) {
+
+        aws_byte_buf_append_dynamic(
+            &meta_request->async_write.buffered_data, &meta_request->async_write.unbuffered_cursor);
+        meta_request->async_write.unbuffered_cursor.len = 0;
+    }
+
+    /* If all unbuffered data is consumed (we sent it, or buffered it) then complete the write's future */
+    if (meta_request->async_write.unbuffered_cursor.len == 0) {
+
+        /* BEGIN CRITICAL SECTION */
+        aws_s3_meta_request_lock_synced_data(meta_request);
+        write_future_to_complete = meta_request->async_write.synced_future;
+        meta_request->async_write.synced_future = NULL;
+        aws_s3_meta_request_unlock_synced_data(meta_request);
+        /* END CRITICAL SECTION */
+
+        if (meta_request->async_write.eof) {
+            *eof = true;
+        }
```

Comment:
trivial: we should move this before we enter the lock and set the `synced_future` to NULL.
Otherwise, the next call is possible to change the eof (I know we haven't set the future result yet, but the protection around the write depends on the future is set or not)

Reply:
Nice catch! I noticed this too when I made the buffer-always change just now